### DISCUSSION TITLE
Example of Dapp doing token approve + transfer

### DISCUSSION THREAD

**Discussion by jmrossy** [wevm/wagmi#2928] [Q&A]
Hi!

Related to a topic I posted a while back around [multiple transactions](https://github.com/wagmi-dev/wagmi/discussions/1564), does anyone know of a good example of sending approve & transfer transactions for a token one after the other?

I've been forming and sending the transactions manually via `sendTransaction` in `@wagmi/core`. This often works but there's an inconsistent bug where Metamask doesn't receive the 2nd sign tx request (i.e. the transfer tx). The modal never pops open.

If there's a better example of running approve and transfer transactions serially I'd be keen to have a look, thanks!

**Comment by avious00** [comment]
@jxom let us know if there's a serial tx example we can look at! thanks in advance

**Comment by avious00** [comment]
@SaadNiaziDev do you by chance know?

**Reply by SaadNiaziDev** [reply]
Will create an example by day end. Hope so it will help you

**Reply by SaadNiaziDev** [reply]
https://github.com/SaadNiaziDev/approve-transaction

clone it and give it a try. I have very weak grip on making react hook so its a basic implementaion here. 
Chain is 97 and you can use any test token for that. If you want specifically token used in this example, send me your address, i will send some for you to test

**Reply by jmrossy** [reply]
Thanks @SaadNiaziDev for the example! However this solution of using two `useContractWrite` hooks [side-by-side](https://github.com/SaadNiaziDev/approve-transaction/blob/main/src/components/approve/index.jsx#L8) would not work for us unfortunately as we would like both txs to be sent from a single onClick handler.

We could modify your solution to grab the hooks' `writeAsync` return callback and then call both in a single click handler, however the callback for the second transaction would have a stale nonce because if was prepared before the first tx was finished. 

This is why we ended up using the `sendTransaction` utility in `@wagmi/core` directly instead of the contract interaction hooks. Manually sending the txs does mostly work except for the bug I mentioned above where it sometimes doesn't pop the Metamask modal for the 2nd tx.

**Reply by SaadNiaziDev** [reply]
Okay. 
1-On which coin you guys are working?
2-There must be error when 2nd transaction doesn't pop even if if you are preparing transactions. 
3- code snippet of this exact part would make it easier to debug.

There is little confusion like why approving when you are going to use senrdTrasaction. Do share a code snippet if possible.

**Reply by Soniachniy** [reply]
Hi @SaadNiaziDev 
I've faced the same issue while trying to buy some ticket from my contract. The flow is simple, user should do the approve action first and call the contract method at the second. But I've the same problem like @jmrossy or @avious00. 

```typescript
async function purchaseTicket(referrer?: string) {
    if (!currentTicketData) return null;

    try {
      const { address: referrerAddress } = checkAddress(
        referrer || EMPTY_STRING
      );

      const approveTokenConfig = await prepareWriteContract({
        ...tokenConfig,
        functionName: 'approve',
        account: address,
        args: [contractWhitelist, currentTicketData.ticketPrice],
      });

      const { hash: approveTokenHash } = await writeContract(
        approveTokenConfig,
      );

      await waitForTransaction({
        hash: approveTokenHash,
      });

      const purchaseTicketConfig = referrerAddress
        ? await prepareWriteContract({
            ...baseConfig,
            functionName: 'purchase_ticket_ref',
            account: address,
            args: [currentTicketData.ticketNumber, referrerAddress as Address],
          })
        : await prepareWriteContract({
            ...baseConfig,
            functionName: 'purchase_ticket',
            account: address,
            args: [currentTicketData.ticketNumber],
          });

      const { hash: purchaseTicketHash } = await writeContract(
        purchaseTicketConfig,
      );

      const receipt = await waitForTransaction({
        hash: purchaseTicketHash,
      });

      ToastService.success({ text: t("HapiId.Toasts.TicketPurchased") });
      return receipt;
    } catch (error) {
      console.error(`Error during ticket purchasing: ${error}`);
      ToastService.error({ text: String(error) });
      return null;
    } finally {
      updateContractData();
    }
  }
``` 
I should say that Metamask dialog is not opened and not redirected at the second time, but if the user try to switch to metamask himself - the second transaction here will appear.

So the behavior like that:
1) user click appove 
2) redirected to MM
3) approve in MM
4) redirected to app
5) wait tx and simulation
6) nothing happened

Supposed: 
6) redirected to MM
7) sign one more tx
8) redirected back to app


What about your code snippet - don't have any chance to add onSuccess method in writeContract since TS alerting

I've "next": "13.4.9", "viem": "1.16.2", "wagmi": "1.4.3", "@web3modal/ethereum": "^2.7.1", "@web3modal/react": "^2.7.1",
Connecting to Sepolia\Mainnet Ethereum via ConnectWallet via Metamask

**Reply by Soniachniy** [reply]
Guys, my problem was resolved by switching to web3modal to v3

**Reply by jmrossy** [reply]
@Solniechniy thanks, are use using just web3modal or web3modal + wagmi?

**Reply by Soniachniy** [reply]
@jmrossy   "@web3modal/wagmi": "^3.1.0",  "viem": "^1.16.3", "wagmi": "^1.4.3" that's my config now

**Comment by avious00** [comment]
hey folks wanted to bump this - we're running into UX issues since last year on this
